src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/LoginHandler.cs

```diff
@@ -0,0 +1,636 @@
+            // Complete the discussion on https://github.com/dotnet/SqlClient/discussions/2689 and then refactor/change this code
+            // to get to the optimization.
+            try
+            {
+                TdsEnums.FeatureExtension requestedFeatures = context.Features.RequestedFeatures;
+                TdsStream stream = context.TdsStream;
+                SqlConnectionEncryptOption encrypt = context.ConnectionOptions.Encrypt;
+                FederatedAuthenticationFeatureExtensionData fedAuthFeatureExtensionData = context.Features.FedAuthFeatureExtensionData;
+                TdsWriter writer = stream.TdsWriter;
+                await writer.WriteIntAsync(length, isAsync, ct).ConfigureAwait(false);
+
+                if (recoverySessionData == null)
+                {
+                    int protocolVersion = encrypt == SqlConnectionEncryptOption.Strict
+                        ? (TdsEnums.TDS8_MAJOR << 24) | (TdsEnums.TDS8_INCREMENT << 16) | TdsEnums.TDS8_MINOR 
+                        : (TdsEnums.SQL2012_MAJOR << 24) | (TdsEnums.SQL2012_INCREMENT << 16) | TdsEnums.SQL2012_MINOR;
+                    await writer.WriteIntAsync(protocolVersion, isAsync, ct).ConfigureAwait(false);
+                }
+                else
+                {
+                    await writer.WriteUnsignedIntAsync(recoverySessionData._tdsVersion, isAsync, ct).ConfigureAwait(false);
+                }
+
+                await writer.WriteIntAsync(context.PacketSize, isAsync, ct).ConfigureAwait(false);
+                await writer.WriteIntAsync(TdsEnums.CLIENT_PROG_VER, isAsync, ct).ConfigureAwait(false);
+                await writer.WriteIntAsync(TdsParserStaticMethods.GetCurrentProcessIdForTdsLoginOnly(), isAsync, ct).ConfigureAwait(false);
+                await writer.WriteIntAsync(0, isAsync, ct).ConfigureAwait(false); // connectionID is unused
+
+                // Log7Flags (DWORD)
+                 
+                int log7Flags = CreateLogin7Flags(context);
+                await writer.WriteIntAsync(log7Flags, isAsync, ct).ConfigureAwait(false);
+                
+                SqlClientEventSource.Log.TryAdvancedTraceEvent("<sc.LoginHandler.WriteLoginData|ADV> TDS Login7 flags = {0}:", log7Flags);
+
+                await writer.WriteIntAsync(0, isAsync, ct).ConfigureAwait(false);  // ClientTimeZone is not used
+                await writer.WriteIntAsync(0, isAsync, ct).ConfigureAwait(false);  // LCID is unused by server
+
+                // Start writing offset and length of variable length portions
+                int offset = TdsEnums.SQL2005_LOG_REC_FIXED_LEN;
+
+                // write offset/length pairs
+
+                // note that you must always set ibHostName since it indicates the beginning of the variable length section of the login record
+                await writer.WriteShortAsync(offset, isAsync, ct).ConfigureAwait(false); // host name offset
```

Comment:
I'd be surprised if this made a difference in the real world, but many of these offsets are flagged as unsigned shorts rather than signed ones.